### ISSUE TITLE
x/build: submit queue, possibly implemented as optional auto-submit on +2 when trybots/slowbots with rebase pass

### ISSUE THREAD

**Issue by minux** [golang/go#12482]
I see a trend of giving +2 together with trybot recently.
because we've defined that +2 means it's ok to submit,
I wonder if trybot could just submit the change (on behave
of the user who gives +2 and also invokes the trybot) if
the trybot runs are successful.

what do you think?

**Comment by bradfitz** [comment]
Russ does not want automatic submission for his own CLs, but he's fine with it for others.

We could add a new checkbox.

**Comment by minux** [comment]
yeah, adding a new checkbox for auto-submit is
fine too (actually I think we can make the 2nd
checkbox do both: invoke the trybot and submit
if it passes.)

**Comment by adg** [comment]
Until we can have the trybots automatically rebase on the HEAD before running, I don't think this is wise. Am I being paranoid?

**Comment by mwhudson** [comment]
On 4 September 2015 at 14:09, Andrew Gerrand <notifications@github.com>
wrote:

> Until we can have the trybots automatically rebase on the HEAD before
> running, I don't think this is wise. Am I being paranoid?
>
I don't think so.

**Comment by minux** [comment]
On Thu, Sep 3, 2015 at 10:09 PM, Andrew Gerrand <notifications@github.com>
wrote:

> Until we can have the trybots automatically rebase on the HEAD before
> running, I don't think this is wise. Am I being paranoid?
>
If that could cause problems, then it will also affect us now because there
is no
way for a human reviewer to notice anything unusual before clicking the
submit
button. The trybot could do a few more checks before submission though.

If we do want to implement, we need to make sure that the trybot will only
try to
submit the revision that the trybot has been run on.

**Comment by bradfitz** [comment]
See also: #9858 (rebase when testing)

**Comment by josharian** [comment]
Here's a slightly more fleshed out suggestion, for discussion.

Add an AutoSubmit score. When set to +1, it requests that GopherBot submit when the following conditions hold.

* The CL has a codereview +2.
* The AutoSubmit score was set to +1 simultaneously to or after the CL received a codereview +2. (This is to ensure that the choice to submit was made after seeing an approved form of the CL.)
* 24 hours have passed since the AutoSubmit score was set, to give others a chance to comment.
* There have been no negative codereview scores since the AutoSubmit score was set to +1 (for the final time).

Once all those have been met, gopherbot does a trybot run *after rebasing*. If it passes, it submits. If it fails, then a human must intervene.

**Comment by josharian** [comment]
cc @dmitshur @bcmills

**Comment by bcmills** [comment]
> * The AutoSubmit score was set to +1 simultaneously to or after the CL received a codereview +2. (This is to ensure that the choice to submit was made after seeing an approved form of the CL.)

I don't think that's necessary: presumably the reviewer can see the AutoSubmit score ahead of time, so if they want to explicitly “+2 with requested changes” they can revoke the AutoSubmit (perhaps by voting AutoSubmit -1?) at that point.

If the uploader of the CL does not have submit permission, then uploading a new snapshot should also revoke the AutoSubmit +1.

**Comment by bcmills** [comment]
> * There have been no negative codereview scores since the AutoSubmit score was set to +1 (for the final time).

This condition seems a bit too strong: a -1 or -2 on the _current_ patchset should certainly cancel auto-submit, but if someone votes -1 or -2 and the CL is subsequently reworked (and those negative votes removed), then it should not be necessary to clear and re-set the AutoSubmit +1 in order for it to take effect.

**Comment by josharian** [comment]
> if they want to explicitly “+2 with requested changes” they can revoke the AutoSubmit

That is true, but I'd rather the default be safe.

> if someone votes -1 or -2 and the CL is subsequently reworked

Again, that is true, but I'd rather be conservative, at least to start.


These minor disagreements notwithstanding, I'd be happy with just about any variant of this.

**Comment by FiloSottile** [comment]
Yup, and I also expected it to be an AutoSubmit vote. I'd like to send simple CLs by `git codereview mail -r gopher@golang.org -trybot -autosubmit` and not have to think about them again, so I'd also relax the "AutoSubmit must come after Code-Review" requirement.

**Comment by matloob** [comment]
cc @andybons @dmitshur 

I would *love*  a submit queue. Ideally we could run all the trybots (including slowbots) when we actually submit a CL. (And if we're more like a submit queue and do a rebase when submitting, the results can be copied straight to build.golang.org when the tests pass and the CL is submitted).

Personally I'd prefer to not have autosubmit but instead have the submit button trigger a test run, submitting if the tests pass, but i'd be pretty happy with autosubmit too.

**Comment by dmitshur** [comment]
CC @rolandshoemaker.

---

Some (small subset of) instances to use as test cases for any future solution:

- https://go.dev/cl/780060 - a semantic conflict resulting in a duplicate import
